### PR TITLE
Implement Fastest CP module

### LINKED ISSUE #27: Fastest CP Widget
Widget at the top of the screen which keeps track of users fastest times on each checkpoint. Unsure if it should be stored in memory or in database.

### PR DESCRIPTION
**Pull request by kdankert**
Started to work a bit on this as an introduction to this project. Already worked a bit with the old version written in PHP and would like to help in this rewrite. 

I based this branch on the manialink branch, as I think, those features will be needed for the widget itself, though as far as I figured out the template engine is not quite ready yet. 

Tests are still missing, but I couldn't figure out the guidelines for tests inside modules, as no module has tests so far. 

In addition to that I still need to verify, that the state of the service is actually thread-safe, as events are most likely to be handled on multiple threads or am I mistaken. 

I thought it would make sense to create a draft pull request though, to let you know that this module is being worked on and to get some early feedback. 

Closes #27.

### CODE REVIEW

**Comment by snixtho** [comment]
Hwy, thanks for helping out! If you'd like, I can add you to the development channel in discord where we are discussing and collaborating on the development of EvoSC#. I think you made a ticket to help out at some point, but was never added to the channel? I can help you get sorted out if you contact me on discord: snixtho#9039

**Comment by kdankert** [comment]
This PR requires #128 to be merged before (target branch needs to be adjusted after).
The last commit splitting the manialink manager interface can either be done in #122 or #128.

**Review by AtomicLiquid — APPROVED** [review]

File: tests/Modules/FastestCp.Tests/FastestCpControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    private readonly FastestCpController _fastestCpController;
+    private readonly Mock<IFastestCpService> _fastestCpService = new();
+
+    public FastestCpControllerTest()
+    {
+        _fastestCpController = new FastestCpController(_fastestCpService.Object);
+    }
+
+    [Fact]
+    public async void ShouldRegisterTimeAndShowWidget()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpControllerTest.cs:20]:
```suggestion
    public async void Should_Register_Time_And_Show_Widget()
```

To keep consistency, we should stick with snake case.

File: tests/Modules/FastestCp.Tests/FastestCpControllerTest.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            IsEndLap = false,
+            IsEndRace = false
+        };
+
+        await _fastestCpController.RegisterCpTime(new object(), waypoint);
+        _fastestCpService.Verify(service => service.RegisterCpTime(waypoint), Times.Once);
+    }
+
+    [Fact]
+    public async void ShouldResetStore()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpControllerTest.cs:44]:
```suggestion
    public async void Should_Reset_Store()

```

File: tests/Modules/FastestCp.Tests/FastestCpManialinkTest.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public FastestCpManialinkTest()
+    {
+        var file = typeof(FastestCpModule).Assembly.GetManifestResourceStream("EvoSC.Modules.Official.FastestCp.Templates.FastestCp.mt");
+        Assert.NotNull(file);
+        var reader = new StreamReader(file);
+        _maniaTemplateEngine.AddTemplateFromString(Key, reader.ReadToEnd());
+    }
+
+    [Fact]
+    public async void ShouldRenderEmptyWidget()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpManialinkTest.cs:21]:
```suggestion
    public async void Should_Render_Empty_Widget()
```

File: tests/Modules/FastestCp.Tests/FastestCpManialinkTest.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        var result = await _maniaTemplateEngine.RenderAsync(Key, new { times = Array.Empty<PlayerCpTime>() },
+            new[] { typeof(FastestCpModule).Assembly });
+        var expected = await File.ReadAllTextAsync("Manialinks/Empty.xml");
+
+        Assert.Equal(result, expected, ignoreLineEndingDifferences: true);
+    }
+
+    [Fact]
+    public async void ShouldRenderFilledWidget()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpManialinkTest.cs:31]:
```suggestion
    public async void Should_Render_Filled_Widget()
```

File: tests/Modules/FastestCp.Tests/FastestCpServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        _manialinkManagerMock
+            .Setup(manager => manager.SendPersistentManialinkAsync(It.IsAny<string>(), It.IsAny<object>()))
+            .Returns(Task.CompletedTask);
+        _fastestCpService = new FastestCpService(_playerManagerServiceMock.Object, _manialinkManagerMock.Object,
+            _loggerFactory, _logger);
+    }
+
+    [Fact]
+    public async void ShouldRegisterTimeAndShowWidget()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpServiceTest.cs:32]:
```suggestion
    public async void Should_Register_Time_And_Show_Widget()
```

File: tests/Modules/FastestCp.Tests/FastestCpServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        _manialinkManagerMock.Verify(
+            manager => manager.SendPersistentManialinkAsync("FastestCp.FastestCp", Capture.In(actualData)), Times.Once);
+        Assert.NotEmpty(actualData);
+        Assert.Equivalent(
+            new { times = new List<PlayerCpTime?> { new("NickName1", 0, TimeSpan.FromMilliseconds(10)) } },
+            actualData[0]);
+    }
+
+    [Fact]
+    public async void ShouldRegisterTimeSlowerTimeAndNotShowWidget()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpServiceTest.cs:67]:
```suggestion
    public async void Should_Register_Time_Slower_Time_And_Not_Show_Widget()
```

File: tests/Modules/FastestCp.Tests/FastestCpServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            IsEndLap = false,
+            IsEndRace = false
+        });
+
+        _manialinkManagerMock.Verify(
+            manager => manager.SendPersistentManialinkAsync("FastestCp.FastestCp", It.IsAny<object>()), Times.Never);
+    }
+
+    [Fact]
+    public async void ShouldResetStore()
```
Comment (snixtho) [inline tests/Modules/FastestCp.Tests/FastestCpServiceTest.cs:94]:
```suggestion
    public async void Should_Reset_Store()
```

**Review by snixtho — CHANGES_REQUESTED** [review]
Nice work! I added some comments and suggestions for you.

**Review by snixtho — APPROVED** [review]
Looks good to me. But it seems like the build and Sonarcloud is failing for not finding the ManiaTemplates project?

**Comment by kdankert** [comment]
Saw that, but couldn't figure out yet, what the problem is. First guess was, that this might be a caching issue, but I'll investigate further.

File: src/Modules/FastestCp/Services/FastestCpService.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            await ShowWidget();
+        }
+    }
+
+    public async Task ResetCpTimesAsync()
+    {
+        _fastestCpStore = GetNewFastestCpStore();
+        await _manialinkManager.HideManialinkAsync("FastestCp.FastestCp");
+        _logger.LogDebug("Hide fastest cp manialink for all users");
```
Comment (AtomicLiquid) [inline src/Modules/FastestCp/Services/FastestCpService.cs:45]:
```suggestion
        _logger.LogDebug("Hiding fastest cp manialink for all users");
```

File: src/Modules/FastestCp/Services/FastestCpService.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    private async Task ShowWidget()
+    {
+        await _manialinkManager.SendPersistentManialinkAsync("FastestCp.FastestCp",
+            new { times = await GetCurrentBestCpTimes() });
+        _logger.LogDebug("Update fastest cp manialink for all users");
```
Comment (AtomicLiquid) [inline src/Modules/FastestCp/Services/FastestCpService.cs:52]:
```suggestion
        _logger.LogDebug("Updating fastest cp manialink for all users");
```

**Review by AtomicLiquid — APPROVED** [review]

**Pull request merged**